vita3k/interface.cpp

```diff
@@ -852,37 +831,42 @@ static std::vector<std::string> split(const std::string &input, const std::strin
     return { first, last };
 }
 
-ExitCode run_app(EmuEnvState &emuenv, Ptr<const void> &entry_point) {
-    const ThreadStatePtr thread = emuenv.kernel.create_thread(emuenv.mem, emuenv.io.title_id.c_str(), entry_point, SCE_KERNEL_DEFAULT_PRIORITY_USER, SCE_KERNEL_THREAD_CPU_AFFINITY_MASK_DEFAULT, static_cast<int>(SCE_KERNEL_STACK_SIZE_USER_MAIN), nullptr);
-    if (!thread) {
+ExitCode run_app(EmuEnvState &emuenv, int32_t main_module_id) {
+    auto entry_point = emuenv.kernel.loaded_modules[main_module_id]->start_entry;
+    auto process_param = emuenv.kernel.process_param.get(emuenv.mem);
+
+    SceInt32 priority = SCE_KERNEL_DEFAULT_PRIORITY_USER;
+    SceInt32 stack_size = SCE_KERNEL_STACK_SIZE_USER_MAIN;
+    SceInt32 affinity = SCE_KERNEL_THREAD_CPU_AFFINITY_MASK_DEFAULT;
+    if (process_param) {
+        auto priority_ptr = Ptr<int32_t>(process_param->main_thread_priority);
+        if (priority_ptr && priority_ptr.valid(emuenv.mem)) {
+            priority = *priority_ptr.get(emuenv.mem);
+        }
+
+        auto stack_size_ptr = Ptr<int32_t>(process_param->main_thread_stacksize);
+        if (stack_size_ptr && stack_size_ptr.valid(emuenv.mem)) {
+            stack_size = *stack_size_ptr.get(emuenv.mem);
+        }
+
+        auto affinity_ptr = Ptr<SceInt32>(process_param->main_thread_cpu_affinity_mask);
+        if (affinity_ptr && affinity_ptr.valid(emuenv.mem)) {
+            affinity = *affinity_ptr.get(emuenv.mem);
```

Comment:
I've never seen invalid pointer or invalid value, but I don't check many modules, so I add this check just in case there is an exact value instead of pointer to value. Hmm. I'm going to add more checks if value itself is valid.

Reply:
No, I mean they shouldn't be necessary. I really doubt the PS Vita kernel checks (or even is able to check) if a pointer is valid or not and would probably cause a crash if an invalid pointer was given. I'd rather have a page fault which can be easily traced than ignoring this error if it ever happens.

Reply:
It's able to check (as any OS with virtual memory), but you are right. There could be something strange for kernel or very old firmware, but users don't run such files.

Reply:
Checks removed